client/core/bookie.go

```diff
@@ -172,52 +184,86 @@ func (c *Core) Sync(host string, base, quote uint32) (*OrderBook, *BookFeed, err
 		return nil, nil, fmt.Errorf("unknown market %s", mkt)
 	}
 
-	// Subscribe
+	obRes, err := dc.Subscribe(base, quote)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	booky = newBookie(func() { dc.StopBook(base, quote) })
+	err = booky.Sync(obRes)
+	if err != nil {
+		return nil, nil, err
+	}
+	dc.books[mkt] = booky
+
+	return booky.book(), booky.feed(), nil
+}
+
+// Subscribe subscribes to the given market's order book via the 'orderbook'
+// request. The response, which includes book's snapshot, is returned. Proper
+// synchronization is required by the caller to ensure that order feed messages
+// aren't processed before they are prepared to handle this subscription.
+func (dc *dexConnection) Subscribe(base, quote uint32) (*msgjson.OrderBook, error) {
+	mkt := marketName(base, quote)
+	// Subscribe via the 'orderbook' request.
+	log.Debugf("Subscribing to the %v order book for %v", mkt, dc.acct.host)
 	req, err := msgjson.NewRequest(dc.NextID(), msgjson.OrderBookRoute, &msgjson.OrderBookSubscription{
 		Base:  base,
 		Quote: quote,
 	})
 	if err != nil {
-		return nil, nil, fmt.Errorf("error encoding 'orderbook' request: %v", err)
+		return nil, fmt.Errorf("error encoding 'orderbook' request: %v", err)
 	}
 	errChan := make(chan error, 1)
 	result := new(msgjson.OrderBook)
 	err = dc.Request(req, func(msg *msgjson.Message) {
 		errChan <- msg.UnmarshalResult(result)
 	})
 	if err != nil {
-		return nil, nil, fmt.Errorf("error subscribing to %s orderbook: %v", mkt, err)
+		return nil, fmt.Errorf("error subscribing to %s orderbook: %v", mkt, err)
 	}
 	err = extractError(errChan, requestTimeout, msgjson.OrderBookRoute)
 	if err != nil {
-		return nil, nil, err
-	}
-
-	booky = newBookie(func() { c.unsub(dc, mkt) })
-	err = booky.Sync(result)
-	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
-	dc.books[mkt] = booky
-
-	return booky.book(), booky.feed(), nil
+	return result, nil
 }
 
-// unsub is the close callback passed to the bookie, and will be called when
+// StopBook is the close callback passed to the bookie, and will be called when
 // there are no more subscribers and the close delay period has expired.
-func (c *Core) unsub(dc *dexConnection, mkt string) {
-	log.Debugf("unsubscribing from %s", mkt)
-
+func (dc *dexConnection) StopBook(base, quote uint32) {
+	mkt := marketName(base, quote)
 	dc.booksMtx.Lock()
-	delete(dc.books, mkt)
-	dc.booksMtx.Unlock()
+	defer dc.booksMtx.Unlock() // hold it locked until unsubscribe request is completed
+
+	// Abort the unsubscribe if feeds exist for the bookie. This can happen if a
+	// bookie's close func is called while a new BookFeed is generated via Sync.
+	if booky, found := dc.books[mkt]; found {
+		booky.mtx.Lock()
+		numFeeds := len(booky.feeds)
+		booky.mtx.Unlock()
+		if numFeeds > 0 {
+			log.Warnf("Aborting booky %p unsubscribe for market %s with active feeds", booky, mkt)
+			return
+		}
```

Comment:
Shouldn't happen with proper use of `dc.booksMtx`, but here to prevent unsubscribing in case active feeds do somehow exist, in which case we need to investigate.

Reply:
The rpc and web servers share the same bookie feeds map, getting their own feed channels.  Also multiple browser windows can open multiple feeds.